### ISSUE TITLE
posessive / lazy modifiers

### ISSUE THREAD

**Issue by matter123** [jeff-hykin/cpp-textmate-grammar#127]
Because of the maximal munch rule [lex.pptoken/3.3](http://eel.is/c%2B%2Bdraft/lex.pptoken#3.3) The grammar needs a way to mark quantifiers as possessive or lazy.

I had some thoughts for what the ai would look like

1.
```ruby
possessive(oneOrMoreOf(/foo/))
```
2.
```ruby
oneOrMoreOf(/foo/).posessive()
```
3.
```ruby
oneOrMoreOf(match: /foo/, :posessive)
```
4.
```ruby
oneOrMoreOfPosessive(/foo/)
```

Advantages:
1 & 2. minimizes code repetition
3. cleanest looking
4. matches the pattern established by lookarounds

Disadvantages:
1 & 2. requires unwrapping the non-capturing group
2. confusing that `.posessive()` applies to the preceding element
3. requires specifying `match:`
4. create a total of 9 functions for qualifiers, or 12 if atomics are supported

@jeff-hykin thoughts?

**Comment by matter123** [comment]
For reference number 2 is
```ruby
    def posessive()
        self_as_string = self.without_default_mode_modifiers
        index = self_as_string[-1] == ')' ? -2 : -1
        if not /[+*?}]/ =~ self_as_string[index]
            raise "\n\n attempted to make a non quantifier pattern posessive\npattern is #{self_as_string}"
        end
        new_regex = /#{self_as_string.insert(index, '+')}/
        
        new_regex.group_attributes = self.group_attributes
        return new_regex
    end
```
number 3 is
```ruby
modifier = ""
if attributes[:type] == nil or attributes[:type] == :greedy
    # pass
elsif attributes[:type] == :posessive
    modifier = "+"
elsif attributes[:type] == :lazy
    modifier = "?"
else
    raise "\n\n argument :type is not :posessive, :greedy :lazy\nregex: #{self.without_default_mode_modifiers}\arguments: #{attributes}"
end
attributes.delete(:type)
```
then interpolating modifer into maybe, oneOrMoreOf, and zeroOrMoreOf's constructed regexps

1 &4 should just be modifications upon those

**Comment by matter123** [comment]
This is blocking some aspects of #135 as `:scope_resolution` can backtrack and cause the lookBehindToAvoids to succeed when it shouldn't.

**Comment by jeff-hykin** [comment]
Sorry, I haven't gotten to this yet. My finals end on Tue so hopefully I'll be able to find more time after they're over.

**Comment by jeff-hykin** [comment]
I'm going to explore this on the plane ride, its a tough choice between them. I'd like there to be an explicit way of doing it, even though right now all of them are implicitly greedy by default. It is probably too wordy for them to all be explicit though.

I'm also going to think about a way to handle repetition ranges like `a{1,3}`

**Comment by matter123** [comment]
For repetition I was thinking:


For exactly 3 times
```ruby
.many(
    count: 3,
)
```
For up to 3 times
```ruby
.many(
    max: 3,
)
```

For 3 or more times
```ruby
.many(
    min: 3,
)
```

For 1 to 3 times
```ruby
.many(
    min: 1,
    max: 3,
)
```

**Comment by jeff-hykin** [comment]
That looks great. My only concern with that and the `oneOrMoreOf(/foo/).posessive()` is that its a bit confusing that some of the `.things` are like a sentence `/thing/.then().then()` but then these (the `.many()` and `.posessive()` are going to be an anti-pattern to that.

**Comment by matter123** [comment]
I should have been more clear, sorry. I was thinking that many follows the way zeroOrMoreOf and maybe do.

Example.
```
ref_deref_pattern = zeroOrMoreOf (/\*/).many(match: /&/, max: 2)
```

**Comment by matter123** [comment]
As far as distinguishing maybe `.posessive!()` to mark that the method modifies the regex "more" than the normal helpers do.

**Comment by matter123** [comment]
Alternatively, functions the modifiy the previous pattern could be an adjative e.g `.posessively`. as in "match zero or more of /foo/, possessively"

**Comment by matter123** [comment]
The following features from Oniguruma Regular Expressions do not have a readable version:
- repetition
  - currently needed
- reluctant/possessive modifiers
  - currently needed
- Last match anchor (\G)
  - currently used
  - not sure it can be made more readable).
- start/end of line anchors (^$)
  - used
  - pretty readable if you know regex
  - not sure $ even works, \n is used everywere instead, might be a reason to have a readable version
- Other anchors (\A, \Z, \z, \b, \B)
  - not really used
- character class
  - used
  - pretty readable, would be hard to make more readable
- comments
  - not needed
- atomic groups
  - could be useful
- subexp calls
  - not used
  - complicated
  - no idea if vscode-textmate supports them

**Comment by jeff-hykin** [comment]
How about we get rid of the `zeroOrMore()` and the `oneOrMore()` and replace all repetitions with this:
```ruby
repeat(
    pattern: /the_pattern/,
    # the following are just for demonstration
    howManyTimes: :asManyAsPossible,
    howManyTimes: :asFewAsPossible,
    howManyTimes: 6, # exact number
    max: 5,
    min: 2,
    keepBacktrackingTillItWorks: false, # possessive = dont backtrack
)
```
We'd assume the `min:` is 1 (e.g. the `+`) unless the user specifically says the `min: 0` in which case we can use the `*`.

To get the `*` in a more readable form we can also do the `maybe(repeat(/blah/))`

**Comment by matter123** [comment]
I would be concerned that, repeat doesn't really seem to clearly map to a regex primitive. Additionally, it feels like patterns become a bunch of `repeat(...).repeat(...).repeat (...)`

**Comment by jeff-hykin** [comment]
Yeah it doesn't clearly map to a regex primitive, but I was actually considering that a helpful abstraction. I'll take a look at what patterns would look like with `repeat()` everywhere before considering it further though.

**Comment by jeff-hykin** [comment]
After replacing the use cases, there are only only 12 `zeroOrMore` and only 2 `oneOrMore` and only a few ranges/possessive/lazy patterns. I don't think the patterns become a bunch of `repeat(...).repeat(...).repeat (...)`, but I see more of the downside with the unclear mapping.

`maybe(repeat(pattern: /thing/))` seems a lot less readable than `zeroOrMoreOf(/thing/)`, and in general `repeat()` is not as english-like as the other ones.

What if instead of repeat, we baked in the repetition to the default pattern.
```
newPattern(
    match: /hello/,
    howManyTimes: :asManyAsPossible, # this converts the pattern to a `*`
    tag_as: "test"
),
newPattern(
    match: /hello/,
    atLeast: 5.times,
    atMost: 2.times,
    tag_as: "test"
)
newPattern(
    match: /hello/,
    atLeast: 0.times,
    tag_as: "test"
)
```

If all of `howManyTimes`, `atLeast`, or `atMost` dont exist, then its a normal pattern
If one of them exists then: 
If atLeast is 0, and atMost is nil then its a `*`
If atLeast is nil, and atMost is nil, then its a `+`
All other cases get custom ranges `{2,5}`
If howManyTimes: :asManyAsPossible, then nothing extra happens
If howManyTimes: :asFewAsPossible, then the `?` is added to the end

**Comment by matter123** [comment]
The first and third option are equivalent right?

**Comment by jeff-hykin** [comment]
I was actually thinking that the first one would have an implicit `at_least: 1.time` rather than `at_least: 0.times`

**Comment by jeff-hykin** [comment]
Oh, my bad I meant `+` not `*`, I'll correct that

**Comment by matter123** [comment]
Well this has the advantage that it can be gradually implented, so we can get some real world examples and see if needs to be modified.

This also looks like it should be a good way to add support for atomic groups

don't backtrack: true, could make the pattern possessive or the group atomic.

I still think oneOrmore and similar should still exist.
`/enum/.then(@std_space).maybe (/class|struct/)` is clearer than `/enum/.then(@std_space).then (match: /class|struct/, atMost: 1.time)`

**Comment by jeff-hykin** [comment]
I wasn't thinking of removing `maybe()`but I think you're right, `oneOrMoreOf` and `zeroOrMoreOf` should probably stay. They're very clear and straightforward. `dont_back_track:` is a good name I think we should go with that, and that would be nice to use it for atomic too.

Lets just add `dont_back_track?`, and a `as_few_as_possible?`, and make the `as_few_as_possible?` only work for the `oneOrMoreOf` and `zeroOrMoreOf` patterns.

Then in the future we can make something like `many()` or `repeat()` that handles more complicated limits

**Comment by matter123** [comment]
Yeah that sounds good.

**Comment by matter123** [comment]
So if I understand correctly, we add a repeat function to Regexp and everything else just aliases to some variant of repeat?

**Comment by jeff-hykin** [comment]
Originally thats what I was thinking, but I think for now we can do something more simple. If you want to go ahead and code that abstraction though you're welcome to, it'll probably make changes later on easier and keep the compiled output consistent

**Comment by matter123** [comment]
The branch `Add/new-regex-patterns` adds support for the option `:dont_back_track?` for quantifiers and atomic groups. Though on line 690: There is the expression `new_regex.group_attributes = self.group_attributes + [ pattern_attributes ] + other_regex.group_attributes` originally this was just `attributes`, even if `pattern_attributes` is just a deep clone of `attributes` the tests fail. @jeff-hykin Do you know why the original `attributes` would work but not the clone?

**Comment by jeff-hykin** [comment]
I'll take a look into the code tomorrow, I'm not sure why without taking a look.

**Comment by matter123** [comment]
Another missing feature is when the start_pattern has a capture group the end pattern can have a back reference to that capture group. See c++ raw string literals.

**Comment by matter123** [comment]
It seems that vscode-textmate does support recursive subexp calls. It seems it creates one capture group for the outermost group. That is, in the example at https://github.com/jeff-hykin/cpp-textmate-grammar/issues/314#issuecomment-512352480, $1 and $0 are equivalent as the named group template captures everything.

As far as a readable syntax, maybe something like.

```
newSubExpression(
    match: /(?<!<)<(?!<)/.oneOrMoreOf(
		match: oneOrMoreOf(match: /[^<>]/, dont_backtrack?: true).maybe(subexpCall("template")),
		dont_backtrack?: true
		).then(/>/),
	reference: "template"
)
